src/subscriber.go

```diff
@@ -104,3 +108,22 @@ func (s *Subscriber) Unsubscribe() {
 
 	s.logger.Info("unsubscribed")
 }
+
+func (s *Subscriber) forwardRTCP() {
+	if s.Track.Kind() != webrtc.RTPCodecTypeVideo {
+		return
+	}
+
+	for {
+		packets, _, err := s.sender.ReadRTCP()
+		if err != nil {
+			if errors.Is(err, io.ErrClosedPipe) || errors.Is(err, io.EOF) {
+				return
+			}
+
+			s.logger.WithError(err).Warn("failed to read RTCP on track")
+		}
+
+		s.publisher.WriteRTCP(packets)
+	}
```

Comment:
General remark: there is a race possible here since both `s.publisher` and `s.sender` may be accessed from different threads (see the `Subscribe()` method in this file), i.e. it's possible to start a goroutine `forwardRTCP()` in `Subscribe()`, then call `Subscribe()` again, which would modify the `s.publisher` and `s.sender` and start `forwardRTCP()` again, which means that 2 go-routines running `forwardRTCP()` would be running. Also it's possible to call `Unsubscribe()` while `forwardRTCP()` is running (since `forwardRTCP()` is running in a separate goroutine) which may lead to a segfault (`Unsubscribe()` sets the `publisher` to `nil`, whereas `forwardRTCP()` continues to access publisher as `s.publisher.WriteRTCP()`.

NB: We don't need to fix it right now in case we merge the refactored version soon (it eliminates races), but I just wanted to mention it, so that we're aware of it 🙂

Reply:
I've noticed that there has been a commit added to address the race condition if I'm not wrong ([latest commit](https://github.com/matrix-org/waterfall/pull/45/commits/8dd122a5b5c6a96959e237f4ea08f4ffe7775cbe)), but unfortunately, it does not fix the race.

The race happens not because you don't check for the `nil`, but because `s.sender` and `s.publisher` may be accessed from 2 threads, but are not protected by a mutex. If a variable is modified in one thread and used in a different thread, it must be protected by a mutex.